nni/contrib/compression/utils/evaluator.py

```diff
@@ -1057,19 +1093,27 @@ def revert_loss(self):
         self.trainer.compute_loss = self._ori_trainer_attr['compute_loss']
 
     def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
-        assert self.trainer.optimizer is not None
-        old_step = self.trainer.optimizer.step
+        def custom_on_train_begin(_, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+            optimizer = self.trainer.deepspeed if self.trainer.deepspeed else self.trainer.callback_handler.optimizer
 
-        def patched_step(_, *args, **kwargs):
-            for task in before_step_tasks:
-                task()
-            # call origin optimizer step method
-            output = old_step(*args, **kwargs)
-            for task in after_step_tasks:
-                task()
-            return output
+            assert optimizer is not None
+            old_step = optimizer.step
+
+            def patched_step(_, *args, **kwargs):
+                for task in before_step_tasks:
+                    task()
+                # call origin optimizer step method
+                output = old_step(*args, **kwargs)
+                for task in after_step_tasks:
+                    task()
+                return output
+
+            optimizer.step = types.MethodType(patched_step, optimizer)
```

Comment:
what does `self.trainer.deepspeed.step` do?

Reply:
it performs processes such as optimizer.step() and scheduler.step().